### PR TITLE
Feature/tensors performance

### PR DESCRIPTION
**Pull request by margarita0303**

### CODE REVIEW

**Comment by grinisrit** [comment]
@margarita0303 please sync with latest `dev`

File: benchmarks/src/jvmMain/kotlin/space/kscience/kmath/benchmarks/SVDBenchmark.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    fun svdPowerMethod(blackhole: Blackhole) {
+        blackhole.consume(
+            tensor.svdPowerMethod()
+        )
+    }
+
+    @Benchmark
+    fun svdGolabKahan(blackhole: Blackhole) {
+        blackhole.consume(
+            tensor.svd()
```
Comment (grinisrit) [inline benchmarks/src/jvmMain/kotlin/space/kscience/kmath/benchmarks/SVDBenchmark.kt:31]: In benchmarks please call `svdGolubKahan`

File: kmath-tensors/src/commonMain/kotlin/space/kscience/kmath/tensors/core/DoubleTensorAlgebra.kt [hunk 2]
```diff
@@ -833,8 +833,33 @@ public open class DoubleTensorAlgebra :
         return qTensor to rTensor
     }
 
-    override fun StructureND<Double>.svd(): Triple<DoubleTensor, DoubleTensor, DoubleTensor> =
-        svd(epsilon = 1e-10)
+    override fun StructureND<Double>.svd(): Triple<DoubleTensor, DoubleTensor, DoubleTensor> {
```
Comment (grinisrit) [inline kmath-tensors/src/commonMain/kotlin/space/kscience/kmath/tensors/core/DoubleTensorAlgebra.kt:836]: you should define a `svdGolubKahan` function and simply call it from the interface `svd`

File: kmath-tensors/src/commonMain/kotlin/space/kscience/kmath/tensors/core/DoubleTensorAlgebra.kt [hunk 3]
```diff
@@ -833,8 +833,33 @@ public open class DoubleTensorAlgebra :
...
+            val curMatrix = DoubleTensor(
+                matrix.shape,
+                matrix.mutableBuffer.array()
+                    .slice(matrix.bufferStart until matrix.bufferStart + matrixSize)
+                    .toDoubleArray()
+            )
+            curMatrix.as2D().svdHelper(uTensors[index].as2D(), sTensorVectors[index], vTensors[index].as2D())
+        }
+
+        return Triple(uTensor.transpose(), sTensor, vTensor)
```
Comment (grinisrit) [inline kmath-tensors/src/commonMain/kotlin/space/kscience/kmath/tensors/core/DoubleTensorAlgebra.kt:861]: Avoid doing a transpose here

File: benchmarks/src/jvmMain/kotlin/space/kscience/kmath/benchmarks/SVDBenchmark.kt [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import kotlinx.benchmark.Scope
+import kotlinx.benchmark.State
+import space.kscience.kmath.tensors.core.BroadcastDoubleTensorAlgebra.svd
+import space.kscience.kmath.tensors.core.DoubleTensorAlgebra
+import space.kscience.kmath.tensors.core.DoubleTensorAlgebra.Companion.svdPowerMethod
+
+@State(Scope.Benchmark)
+class SVDBenchmark {
+    companion object {
+        val tensor = DoubleTensorAlgebra.randomNormal(intArrayOf(10, 10, 10), 0)
```
Comment (grinisrit) [inline benchmarks/src/jvmMain/kotlin/space/kscience/kmath/benchmarks/SVDBenchmark.kt:18]: please run several benchmarks small - (5,5) , medium (10,10) - large (50,50) - very large (100,100). In the initialisation assert that norm(A - U S Vt) < 1e-9

File: kmath-tensors/src/commonMain/kotlin/space/kscience/kmath/tensors/core/DoubleTensorAlgebra.kt (hunk 2 above)
Comment (grinisrit) [reply kmath-tensors/src/commonMain/kotlin/space/kscience/kmath/tensors/core/DoubleTensorAlgebra.kt:836]: And `svdGolubKahan` should have all the parameters with default values for the interface call with `svd`

File: kmath-tensors/src/commonMain/kotlin/space/kscience/kmath/tensors/core/internal/linUtils.kt [hunk 5]
```diff
@@ -349,3 +350,294 @@ internal fun DoubleTensorAlgebra.svdHelper(
...
+                        }
+                    }
+                }
+                for (k in i until m) {
+                    this[k, i] = this[k, i] * scale
+                }
+            }
+        }
+
+        w.mutableBuffer.array()[w.bufferStart + i] = scale * g
```
Comment (grinisrit) [inline kmath-tensors/src/commonMain/kotlin/space/kscience/kmath/tensors/core/internal/linUtils.kt:429]: Use `mutableBuffer`setter method - do that in other places as well in the library.

File: benchmarks/src/jvmMain/kotlin/space/kscience/kmath/benchmarks/SVDBenchmark.kt (hunk 1 above)
Comment (margarita0303) [reply benchmarks/src/jvmMain/kotlin/space/kscience/kmath/benchmarks/SVDBenchmark.kt]: ok

File: kmath-tensors/src/commonMain/kotlin/space/kscience/kmath/tensors/core/DoubleTensorAlgebra.kt (hunk 2 above)
Comment (margarita0303) [reply kmath-tensors/src/commonMain/kotlin/space/kscience/kmath/tensors/core/DoubleTensorAlgebra.kt:836]: 1) ok 2) added parameters iterations and epsilon

File: kmath-tensors/src/commonMain/kotlin/space/kscience/kmath/tensors/core/internal/linUtils.kt (hunk 5 above)
Comment (margarita0303) [reply kmath-tensors/src/commonMain/kotlin/space/kscience/kmath/tensors/core/internal/linUtils.kt]: changed in section kmath-tensors

File: kmath-tensors/src/commonMain/kotlin/space/kscience/kmath/tensors/core/DoubleTensorAlgebra.kt (hunk 3 above)
Comment (margarita0303) [reply kmath-tensors/src/commonMain/kotlin/space/kscience/kmath/tensors/core/DoubleTensorAlgebra.kt]: done

**Pull request closed without merging**